crates/core/src/protobuf.rs

```diff
@@ -34,9 +38,133 @@ pub mod node_ctl_svc {
+    }
+
+    impl MetadataStoreProxy {
+        pub fn new(channel: Channel) -> Self {
+            Self {
+                client: Arc::new(Mutex::new(
+                    MetadataProxySvcClient::new(channel)
+                        .send_compressed(CompressionEncoding::Gzip)
+                        .accept_compressed(CompressionEncoding::Gzip),
+                )),
+            }
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl MetadataStore for MetadataStoreProxy {
+        async fn get(&self, key: ByteString) -> Result<Option<VersionedValue>, ReadError> {
+            let mut client = self.client.lock().await;
+
+            let request = GetRequest {
+                key: key.to_string(),
+            };
+
+            let response = client.get(request).await.map_err(to_read_err)?.into_inner();
+
+            response
+                .value
+                .map(TryInto::try_into)
+                .transpose()
+                .map_err(ReadError::terminal)
+        }
+
+        async fn get_version(&self, key: ByteString) -> Result<Option<Version>, ReadError> {
+            let mut client = self.client.lock().await;
+
+            let request = GetRequest {
+                key: key.to_string(),
+            };
+
+            let response = client
+                .get_version(request)
+                .await
+                .map_err(to_read_err)?
+                .into_inner();
+
+            response
+                .version
+                .map(TryInto::try_into)
+                .transpose()
+                .map_err(ReadError::terminal)
+        }
+
+        async fn put(
+            &self,
+            key: ByteString,
+            value: VersionedValue,
+            precondition: Precondition,
+        ) -> Result<(), WriteError> {
+            let mut client = self.client.lock().await;
+
+            let request = PutRequest {
+                key: key.to_string(),
+                value: Some(value.into()),
+                precondition: Some(precondition.into()),
+            };
+
+            client.put(request).await.map_err(to_write_err)?;
+
+            Ok(())
+        }
+
+        async fn delete(
+            &self,
+            key: ByteString,
+            precondition: Precondition,
+        ) -> Result<(), WriteError> {
+            let mut client = self.client.lock().await;
+
+            let request = DeleteRequest {
+                key: key.to_string(),
+                precondition: Some(precondition.into()),
+            };
+
+            client.delete(request).await.map_err(to_write_err)?;
+
+            Ok(())
+        }
+
+        async fn provision(&self, _: &NodesConfiguration) -> Result<bool, ProvisionError> {
+            Err(ProvisionError::NotSupported(
+                "Provision not supported over metadata proxy".to_owned(),
+            ))
+        }
+    }
+}
+
+fn to_read_err(status: Status) -> ReadError {
+    ReadError::terminal(status)
```

Comment:
Maybe leave a comment why it's ok to treat all statuses as terminal for the time being.